### PR TITLE
Moving general-purpose rapply from Programs to Init

### PR DESCRIPTION
**Pull request by herbelin**
Who has experience with `rapply`? Shouldn't it be at some time the default `apply` (or, at least, unifall would lead to identify apply and rapply, right?)?

There is an alternative definition of `rapply` in TLC (and Software Foundations) which seems currently broken. Since there is a version of `rapply` already in Coq, maybe should we just rely on this one??

- [ ] Added **changelog** (?)
- [x] Updated **documentation**.

### CODE REVIEW

**Comment by andres-erbsen** [comment]
I believe `rapply` and `apply` also differ in elaboration order and handling of `/\` in hypotheses (and maybe lemmas). @JasonGross probably knows

**Review by JasonGross — APPROVED** [review]
I'm in favor.

I agree that the implementation of `apply` should be moved to the `refine` engine (ideally while keeping the current behavior of decomposing single-constructor inductives).  However, the distinction between `apply` and `eapply` should be kept, and the move should be done in OCaml, not Ltac, I think.

**Comment by herbelin** [comment]
I realize that Program's `rapply` tries maximal application first, so that it may encounters limits of unification and fail on non atomic goals with flexible head variables, as in:
```coq
Require Import Program.
Axiom A : forall R:nat -> Prop,  forall x, R x.
Goal forall R:nat -> Prop,  forall x, R x.
Fail rapply A. (* Unable to unify "?Goal ?Goal0" with "forall (R : nat -> Prop) (x : nat), R x" *)
```
I guess that the motivation to start with maximal application is efficiency, so I don't know what to think.

This can be addressed if moved to OCaml though.

**Comment by Alizter** [comment]
I would prefer an OCaml implementation for rapply since we are paying attention to it.

**Comment by herbelin** [comment]
> I would prefer an OCaml implementation for rapply

I made an experimental branch at https://github.com/herbelin/github-coq/pull/new/master+native-rapply so that we can discuss what to do with it, based on concrete code.

It provides `rapply` and `erapply` and supports destructing single-constructor. The `rapply in` and `erapply in` are still to be done though.

**Comment by gares** [comment]
@coqbot run full CI

**Comment by SkySkimmer** [comment]
@coq/stdlib-maintainers this has been stalled for a while, please take a decision (eg merge / close / say that something specific needs to be done before merge)

If you merge before the branch you can put back the milestone.

**Comment by SkySkimmer** [comment]
ping @coq/stdlib-maintainers

**Comment by andres-erbsen** [comment]
I am in favor of merging one implementation or the other. The Ltac one looks good to me despite limitations; if we want the OCaml one somebody else should review in detail. I don't have a particular for it being in prelude, so if those opposing that placement could offer an alternative that might be better indeed.

**Review by JasonGross — APPROVED** [review]
I am a fan of moving this to Init.  I think probably an OCaml variant might be better, but I'm happy to have this merged and defer the decision on the OCaml implementation to later.

**Comment by andres-erbsen** [comment]
Hearing no suggested changes, I intend to merge this PR next time I look at it.

**Comment by andres-erbsen** [comment]
@coqbot run full ci

it's been almost a year

**Comment by andres-erbsen** [comment]
@coqbot merge now

**Pull request merged**
